README.md
# Tetris
Peli jossa liikutetaan tippuvia palikoita ja kerätään pisteitä muodostamalla niistä vaakasuoria rivejä kentän pohjalle.

## Käyttöohjeet
1. Asenna riippuvuudet komennolla: 
```bash 
poetry install
```
2. Käynnistä peli komennolla: 
```bash
poetry run invoke start
```
### Miten peliä pelataan:
- Pelissä palikoita voidaan liikutella oikealle tai vasemmalle nuolinäppäimillä.
- Kun painaa ylös -nuolinäppäintä, palikka kääntyy.
- Alas -nuolinäppäintä pohjaan painamalla palikka tippuu nopeammin.
- Painamalla x -näppäintä palikan voi vaihtaa talteen "hold" -kenttään.
- Game over -ruudussa halutun vaihtoehdon voi vahvistaa painamalla enter -näppäintä.

### Miten peliä testataan:
Peliä testataan komennolla:
```bash
poetry run invoke test
```
Testikattavuusraportin saa komennolla:
```bash
poetry run invoke coverage-report
```
Pylint tarkistukset voi suorittaa komennolla:
```bash
poetry run invoke lint
```
## Dokumentaatio
- [Vaatimusmäärittely](./dokumentaatio/vaatimusmaarittely.md)
- [Arkkitehtuurikuvaus](./dokumentaatio/arkkitehtuuri.md)
- [Työaikakirjanpito](./dokumentaatio/tuntikirjanpito.md)
- [Changelog](./dokumentaatio/changelog.md)


dokumentaatio/changelog.md
## Viikko 4
- Palikoiden ja tason luokat suunniteltu
- Pelin avatessa kenttä muodostuu ruudulle

## Viikko 5
- Palikat muodostuvat kentälle oikein
- Palikat tippuvat kentän pohjalle ja pysähtyvät

## Viikko 6
- Pelaaja voi liikuttaa palikoita kentässä
- Pelaaja voi käännellä palikoita
- Palikat pysähtyvät osuessaan toisiin pysähtyneisiin palikoihin

## Viikko 7
- Palikat tippuvat nopeampaa vauhtia pelin edetessä
- Pelaaja voi tiputtaa palikoita nopeammin painavalla alas -nuolinäppäintä.
- Pelin ulkonäkyä on parannettu
- Palikoita voi vaihtaa vara palikoiksi painamalla enter:iä
- Pelissä on game over -ruutu
- Pelissä voi ansaita pisteitä ja aiempien pelikertojen ennätykset näkyvät ruudulla.


src/tests/block_test.py
import pygame
import unittest

from stage import Stage

class Test_Block(unittest.TestCase):
    def setUp(self):
        self.stage = Stage()
        self.stage.map = [[0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,1,0,0,0,0,0],
        [0,0,0,1,1,1,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0]]



    def test_blocks_fall_correctly(self):
        correct = [[0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,1,0,0,0,0,0],
        [0,0,0,1,1,1,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0]]

        self.stage.drop_block()

        self.assertEqual(correct, self.stage.map)
    
    def test_freeze_correctly(self):
        test1 = Stage()
        test1.map = [[0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [1,1,1,1,1,1,1,1,1,1],
        [1,1,1,1,1,1,1,1,1,1]]

        test1.drop_block()

        test_map_correct = [
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0]]

        self.assertEqual(test1.map[17:19], test_map_correct)

    
    def test_blocks_moves_correctly(self):
        correct = [[0,0,0,0,0,0,0,0,0,0],
        [0,0,0,1,0,0,0,0,0,0],
        [0,0,1,1,1,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0]]

        self.stage.move_block("L")

        self.assertEqual(correct, self.stage.map)

        correct2 =[[0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,1,0,0,0,0,0],
        [0,0,0,1,1,1,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0]]

        self.stage.move_block("R")
        self.assertEqual(correct2, self.stage.map)

laskarit/viikko2/maksukortti/src/tests/maksukortti_test.py
import unittest
from maksukortti import Maksukortti

class TestMaksukortti(unittest.TestCase):
    def setUp(self):
        self.kortti = Maksukortti(1000)

    def test_konstruktori_asettaa_saldon_oikein(self):
        self.assertEqual(str(self.kortti), "Kortilla on rahaa 10.00 euroa")

    def test_syo_edullisesti_vahentaa_saldoa_oikein(self):
        self.kortti.syo_edullisesti()

        self.assertEqual(str(self.kortti), "Kortilla on rahaa 7.50 euroa")

    def test_syo_maukkaasti_vahentaa_saldoa_oikein(self):
        self.kortti.syo_maukkaasti()

        self.assertEqual(str(self.kortti), "Kortilla on rahaa 6.00 euroa")

    def test_syo_edullisesti_ei_vie_saldoa_negatiiviseksi(self):
        kortti = Maksukortti(200)
        kortti.syo_edullisesti()

        self.assertEqual(str(kortti), "Kortilla on rahaa 2.00 euroa")

    def test_kortille_voi_ladata_rahaa(self):
        self.kortti.lataa_rahaa(2500)

        self.assertEqual(str(self.kortti), "Kortilla on rahaa 35.00 euroa")

    def test_kortin_saldo_ei_ylita_maksimiarvoa(self):
        self.kortti.lataa_rahaa(20000)

        self.assertEqual(str(self.kortti), "Kortilla on rahaa 150.00 euroa")

    def test_syo_maukkaasti_ei_vie_saldoa_negatiiviseksi(self):
        self.kortti = Maksukortti(300)
        self.kortti.syo_maukkaasti()

        self.assertEqual(str(self.kortti), "Kortilla on rahaa 3.00 euroa")

    def test_negatiivinen_ei_muuta_korttia(self):
        self.kortti.lataa_rahaa(-200)

        self.assertEqual(str(self.kortti), "Kortilla on rahaa 10.00 euroa")

    def test_edullinen_lounas_minimi(self):
        self.kortti = Maksukortti(250)
        self.kortti.syo_edullisesti()

        self.assertEqual(str(self.kortti), "Kortilla on rahaa 0.00 euroa")

    def test_maukas_lounas_minimi(self):
        self.kortti = Maksukortti(400)
        self.kortti.syo_maukkaasti()

        self.assertEqual(str(self.kortti), "Kortilla on rahaa 0.00 euroa")



src/stage.py
import pygame
from blocks import Blocks

class Stage():
    """Luokka jollaa ylläpidetään pelikenttän tilaa.

    Attributes:
        map: Kartoitus kentästä. Kentässä 0 on tyhjä tila, 1 on palikka ja 2 on pysähtynyt palikka.
        
        rotation_b: eli rotation block, jota käytetään palikan sijainnin muistamiseen kentässä,
        ja palikan kääntämiseen.
        
        rotations: Sisältää palikan eri asennot ja tiedon sen hetkisestä asennosta.

     """
    def __init__(self):
        self.map = [[0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0]]

        self.game_over = False
        self.block = Blocks()
        self.color1 = None
        self.color2 = None
    
    def add_new_block(self):
        """Lisää uuden palikan kenttään."""
        block1 = self.block.generate_random_block()
        block = block1[0]
        self.color1 = block1[1]
        error = self.error_check((0,8),block[0])
        if error:
            self.game_over = True
        for y in range(0,4):
            for x in range(0,5):
                if block[0][y][x] == 3:
                    self.block.rotation_b = (0,(y, (x+5+3)))
                else:
                    if error != True:
                        self.map[y][x+3] = block[0][y][x]
        self.block.rotations = block
    
    def switch_block(self):
        """Vaihtaa palikan varapalikkaa."""
        if self.block.spare == []:
            self.block.spare = self.block.rotations
            self.color2 = self.color1
            self.clear_stage(self.map)
            self.add_new_block()

        else:
            old_block = self.block.rotations
            new_block = self.block.spare
            rotationb = self.block.rotation_b
            old_color = self.color1
            if self.block.rotation_b[1][1] > 10:
                rotationb = (self.block.rotation_b[0], (self.block.rotation_b[1][0],10))
            if self.block.rotation_b[1][0] > 16:
                return
            error = self.error_check(rotationb[1],new_block[0])
            if error:
                return

            self.color1 = self.color2
            self.color2 = old_color
            self.block.spare = old_block
            self.block.rotations = new_block
            self.block.rotation_b = rotationb
            self.clear_stage(self.map)
            self.insert_block(self.block.rotation_b[1],self.block.rotations[0])
            if rotationb[1][1] == 10:
                self.move_block("R")


    def insert_block(self,rotation_b,block):
        """Sijoittaa palikan kenttään.
        Args:
            rotation_b: sisältää palikan koordinaatin.
            block: palikka listana."""
        move = False
        y2 = -1
        x2 = -1
        yb = rotation_b[0]
        xb = rotation_b[1]-5
        if xb < 0:
            xb = 0
            move = True
        for y in range(yb,(yb+4)):
            y2 += 1
            for x in range(xb,xb+5):
                x2 += 1
                if self.map[y][x] == 2:
                    continue
                if block[y2][x2] == 3:
                    continue
                self.map[y][x] = block[y2][x2]
            x2 = -1
            
        if move:
            self.move_block("L")
    
    def drop_block(self):
        """Tiputtaa palikkaa.
        Returns:
            score: Pisteet joita saadaan jos tiputettu palikka jäätyy."""
        for y in range(19,-1,-1):
            for x in range(0,10):
                if self.map[y][x] == 1:
                    if y == 19 or self.map[y+1][x] == 2:
                        score = self.freeze_block()
                        self.add_new_block()
                        return score

        for y in range(19,-1,-1):
            for x in range(0,10):
                if self.map[y][x] == 1:
                    self.map[y][x] = 0
                    self.map[y+1][x] = 1
        self.block.move_rotation_b("D")
    
    def move_block(self, direction):
        """Liikuttaa palikkaa joko oikealle tai vasemmalle.
        Args:
            direction: Siirron suunta, joko "R" tai "L".
        """
        if direction == "R":
            for y in range(0,20):
                if self.map[y][9] == 1:
                    return
            
            prev = 0
            for row in self.map:
                for x in row:
                    if prev == 1 and x == 2:
                        return
                    prev = x

            for y in range(0,20):
                for x in range(9,-1,-1):
                    if self.map[y][x] == 1:
                        self.map[y][x] = 0
                        self.map[y][x+1] = 1
            self.block.move_rotation_b("R")

        elif direction == "L":
            for y in range(0,20):
                if self.map[y][0] == 1:
                    return
            
            prev = 0
            for row in self.map:
                for x in row:
                    if prev == 2 and x == 1:
                        return
                    prev = x

            for y in range(0,20):
                for x in range(0,10):
                    if self.map[y][x] == 1:
                        self.map[y][x] = 0
                        self.map[y][x-1] = 1
            self.block.move_rotation_b("L")          

    def freeze_block(self):
        """Pysäyttää palikan paikoilleen.
        Returns:
            self.check_rows: palauttaa funktion check_rows laskemat pisteet."""
        for y in range(0,20):
            for x in range(0,10):
                if self.map[y][x] == 1:
                    self.map[y][x] = 2
        return self.check_rows()
    
    def check_rows(self):
        """Tarkistaa onko kentässä rivejä jotka ovat täynä.
        Returns:
            score: palauttaa saatujen pisteiden määrän.
        """
        score = 1
        for y in range(19,0,-1):
            while self.map[y] == [2,2,2,2,2,2,2,2,2,2]:
                score += 10
                for y2 in range(y,1,-1):
                    prev = self.map[y2-1][:]
                    self.map[y2] = prev
        return score
    
    def clear_stage(self,map1):
        """Tyhjentää kentän siinä olevasta palikasta.
        Args:
            map1: kenttä josta palikka poistetaan."""
        for y in range(0,20):
            for x in range(0,10):
                if map1[y][x] == 1:
                    map1[y][x] = 0
        return
        
    def rotate_block(self):
        """Kääntää palikkaa."""
        block_y = self.block.rotation_b[1][0]
        block_x = self.block.rotation_b[1][1] -5
        move = False
        old = self.block.rotation_b
        if block_x < 0:
            block_x = 0
            self.block.rotation_b = (self.block.rotation_b[0], (self.block.rotation_b[1][0],0))
            move = True
            dir = "L"    
        if block_x > 5:
            block_x = 5
            self.block.rotation_b = (self.block.rotation_b[0], (self.block.rotation_b[1][0],10))
            move = True
            dir = "R"
        if block_y > 16:
            self.block.rotation_b = old
            return
        if self.block.rotation_b[0]+1 < 4:
            r = self.block.rotation_b[0]+1
            self.block.rotation_b = ((self.block.rotation_b[0]+1), self.block.rotation_b[1])
        else:
            r = 0
            self.block.rotation_b = (0, self.block.rotation_b[1])

        next = self.block.rotations[r]
        error = self.error_check(self.block.rotation_b[1],next)
        if error:
            self.block.rotation_b = old
            return
        
        self.clear_stage(self.map)
        self.insert_block(self.block.rotation_b[1], next)

        if move:
            self.move_block(dir)

    def error_check(self,rotation_b,block):
        """Tarkistaa voiko palikkaa lisätä kenttään tiettyyn kohtaan.
        Args:
            rotation_b: palikan sijoittamiseen käytettävä koordinaatti.
            block: palikka.
        Returns:
            True tai False, riippuen siitä voiko palikkaa sijoittaa vai ei."""
        check_map = []
        over = 0
        for row in self.map:
            check_map.append(row[:])
        self.clear_stage(check_map)

        y2 = -1
        x2 = -1
        if rotation_b[1] == 0:
            rotation_b = (rotation_b[0], 5)
        if rotation_b[1]-5 < 0:
            over = -(rotation_b[1] - 5)
        for y in range(0,4):
            for x in range(0,over):
                if block[y][x] == 1:
                    return True

        for y in range(rotation_b[0],(rotation_b[0]+4)):
            y2 = y2+1
            for x in range(rotation_b[1]-5+over,(rotation_b[1])):
                x2 = x2+1
                if check_map[y][x] == 2 and block[y2][x2] == 1:
                    return True
            x2 = -1
        return False
    
    def retry(self):
        self.map = [[0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0,0,0]]
        self.block = Blocks()
        self.color1 = None
        self.color2 = None
        self.add_new_block()

src/blocks.py
import pygame
import random

class Blocks():
    """Luokka, jossa on kaikki eri palikoiden muodot nimetty: I,T,Z,S,L,J ja O.
    Attributes:
            spare: Lista joka sisältää vara palikan.

            rotation_b: Eli rotation block, jota käytetään palikan sijainnin muistamiseen kentässä,
            ja palikan kääntämiseen.
        
            rotations: Sisältää palikan eri asennot ja tiedon sen hetkisestä asennosta.

            spare: Vara-palikan lista.
            """
    def __init__(self):
        self.I =[[[3,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0]],
        [[3,0,0,0,0],
        [0,1,1,1,1],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0]],
        [[3,0,0,0,0],
        [0,1,1,1,1],
        [0,0,0,0,0],
        [0,0,0,0,0]]]

        self.T =[[[3,0,1,0,0],
        [0,1,1,1,0],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,0,1,0,0],
        [0,0,1,1,0],
        [0,0,1,0,0],
        [0,0,0,0,0]],
        [[3,0,0,0,0],
        [0,1,1,1,0],
        [0,0,1,0,0],
        [0,0,0,0,0]],
        [[3,0,1,0,0],
        [0,1,1,0,0],
        [0,0,1,0,0],
        [0,0,0,0,0]]]

        self.Z = [[[3,1,1,0,0],
        [0,0,1,1,0],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,0,1,0,0],
        [0,1,1,0,0],
        [0,1,0,0,0],
        [0,0,0,0,0]],
        [[3,1,1,0,0],
        [0,0,1,1,0],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,0,1,0,0],
        [0,1,1,0,0],
        [0,1,0,0,0],
        [0,0,0,0,0]]]

        self.S =[[[3,0,1,1,0],
        [0,1,1,0,0],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,0,1,0,0],
        [0,0,1,1,0],
        [0,0,0,1,0],
        [0,0,0,0,0]],
        [[3,0,1,1,0],
        [0,1,1,0,0],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,0,1,0,0],
        [0,0,1,1,0],
        [0,0,0,1,0],
        [0,0,0,0,0]]]

        self.L = [[[3,0,1,0,0],
        [0,0,1,0,0],
        [0,0,1,1,0],
        [0,0,0,0,0]],
        [[3,0,0,0,0],
        [0,1,1,1,0],
        [0,1,0,0,0],
        [0,0,0,0,0]],
        [[3,1,1,0,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,0,0,0]],
        [[3,0,0,1,0],
        [0,1,1,1,0],
        [0,0,0,0,0],
        [0,0,0,0,0]]]

        self.J = [[[3,0,1,0,0],
        [0,0,1,0,0],
        [0,1,1,0,0],
        [0,0,0,0,0]],
        [[3,1,0,0,0],
        [0,1,1,1,0],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,0,1,1,0],
        [0,0,1,0,0],
        [0,0,1,0,0],
        [0,0,0,0,0]],
        [[3,0,0,0,0],
        [0,1,1,1,0],
        [0,0,0,1,0],
        [0,0,0,0,0]]]

        self.O = [[[3,1,1,0,0],
        [0,1,1,0,0],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,1,1,0,0],
        [0,1,1,0,0],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,1,1,0,0],
        [0,1,1,0,0],
        [0,0,0,0,0],
        [0,0,0,0,0]],
        [[3,1,1,0,0],
        [0,1,1,0,0],
        [0,0,0,0,0],
        [0,0,0,0,0]]]

        self.spare = []
        self.rotation_b = (0, (0,0))
        self.rotations = []        
    
    def generate_random_block(self):
        """Tuottaa uuden palikan valitsemalla sen satunnaisesti listasta.
        Returns:
            new_block: Palikka, jonka muodot on matriiseina listassa.
        """
        list = [(self.I, "cyan"), (self.J, "blue"), (self.L, "orange"), (self.S, "green"), (self.T, "purple"), (self.Z, "red"), (self.O, "yellow")]
        pick = random.randint(0,6)
        new_block = list[pick]
        return new_block

    def move_rotation_b(self, dir):
        """Liikuttaa kääntämiseen käytettävää tilaa oikealle, vasemmalle tai alas. 
        Args:
            dir: Siirron suunta, joko "R", "L" tai "D".
        """
        if dir == "D":
            old = self.rotation_b
            self.rotation_b = (old[0], ((old[1][0]+1),old[1][1]))
            return
        
        if dir == "L":
            x = -1
        elif dir == "R":
            x = 1

        old = self.rotation_b
        self.rotation_b = (old[0], (old[1][0],(old[1][1]+x)))


src/renderer.py
import pygame
import os

dirname = os.path.dirname(__file__)

class Renderer():
    """Luokka, joka renderöi pelikentän.
    
    Args:
        stage: Pelikenttä.
        stage.lenght: Pelikentän korkeus.
        stage.widght: Pelikentän leveys.
        display: Ruutu jolle kenttä piirretään.
        block1: Palikan väri.
        block2: Vara-palikan väri.
        font: Pelissä käytettävä teksti fontti.
        block3: Jäätyneiden palikoiden väri.
    """
    def __init__(self, stage, display):
        self.stage = stage
        self.stage_lenght = len(self.stage.map)
        self.stage_width = len(self.stage.map[0])
        self.display = display
        self.block1 = None
        self.block2 = None
        self.font = pygame.font.SysFont("arial", 30)
        self.block3 = pygame.image.load(os.path.join(dirname, "assets", "gray.jpg"))
    
    def render_game(self):
        """Renderöi kerralla kentän ja vara-palikan."""
        self.render_stage()
        self.render_spare()
        pygame.display.update()

    def block_color(self):
        """Selvittää palikoiden värit."""
        color1 = self.stage.color1
        color1 = color1 + ".jpg"
        self.block1 = pygame.image.load(os.path.join(dirname, "assets", color1))

        if self.stage.color2 != None:
            color2 = self.stage.color2
            color2 = color2 + ".jpg"
            self.block2 = pygame.image.load(os.path.join(dirname, "assets", color2))

    def render_stage(self):
        """Renderöi pelikentän."""
        self.block_color()
        pygame.draw.rect(self.display,(211,211,211),(188,48,(30*10+4),(30*20+4)),2)
        for y in range(0,self.stage_lenght):
            for x in range(0,self.stage_width):
                if self.stage.map[y][x] == 0:
                    pygame.draw.rect(self.display,(55,55,55),(190+x*30,50+y*30,30,30),1)
                elif self.stage.map[y][x] == 1:
                    self.display.blit(self.block1, (190+x*30,50+y*30))
                elif self.stage.map[y][x] == 2:
                    self.display.blit(self.block3, (190+x*30,50+y*30))
    
    def render_spare(self):
        """Renderöi vara-palikan ruudulle."""
        hold = self.font.render("Hold (x)", True, (255,255,255))
        self.display.blit(hold, (555,20))
        spare = self.stage.block.spare
        pygame.draw.rect(self.display,(211,211,211),(550,60,(30*5),(30*5)),2)
        if spare == []:
            return
        spare = self.stage.block.spare[0]
        for y in range(0,4):
            for x in range(0,5):
                if spare[y][x] == 1:
                    self.display.blit(self.block2, (560+30*x,75+30*y))
    
    def render_info(self, score, max_score, level):
        """Renderöi tekstinä tietoa ruudulle pelistä.
        Args:
            score: Tämän hetkisen pelin pisteet.
            max_score: Peli kertojen suurin pistemäärä.
            level: Pelin nopeus taso.
            """
        note1 = self.font.render(("Score: "+str(score)), True, (255,255,255))
        note2 = self.font.render(("Best score: "+str(max_score)), True, (255,255,255))
        note3 = self.font.render(("Level: "+str(level)), True, (255,255,255))
        self.display.blit(note1, (555,400))
        self.display.blit(note2, (555,450))
        self.display.blit(note3, (555,300))
    
    def game_over(self, option):
        """Pelin päätettyä näytettävien vaihtoehtojen renderöinti.
        Args:
            option: Tämän hetkinen vaitoehto, yes tai no.
            """
        pygame.draw.rect(self.display,(0,0,0),(400,400,600,600))
        if option == "yes":
            color1 = ((255,255,255))
            color2 = ((150,150,150))
        elif option == "no":
            color1 = ((150,150,150))
            color2 = ((255,255,255))
        note1 = self.font.render(("Retry?"), True, (255,255,255))
        note2 = self.font.render(("Yes"), True, (color1))
        note3 = self.font.render(("No"), True, (color2))
        self.display.blit(note1, (450,400))
        self.display.blit(note2, (450,450))
        self.display.blit(note3, (450,500))
        pygame.display.update()
    
        
        




        


src/game_loop.py
import pygame

class GameLoop():
    """Luokka, joka lukee pelaajan syötteen ja joko tekee muutoksia pelikenttään tai lopetetaan peli.
    
    Attributes:
    stage: Luokka saa pelikentän johon muutokset tehdään."""
    def __init__(self, stage):
        self.stage = stage
        self.drop = False

    def events(self):
        """Käy läpi pelaajaan syötteen."""
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_RIGHT:
                    self.stage.move_block("R")
                if event.key == pygame.K_LEFT:
                    self.stage.move_block("L")
                if event.key == pygame.K_UP:
                    self.stage.rotate_block()
                if event.key == pygame.K_DOWN:
                    self.drop = True
                if event.key == pygame.K_x:
                    self.stage.switch_block()
            if event.type == pygame.KEYUP:
                if event.key == pygame.K_DOWN:
                    self.drop = False
    
    def game_over(self):
        """Game over -näytössä tehtävät valinnat."""
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_DOWN:
                    return "no"
                if event.key == pygame.K_UP:
                    return "yes"
                if event.key == pygame.K_RETURN:
                    return True

src/index.py
import pygame
from stage import Stage
from clock import Clock
from renderer import Renderer
from game_loop import GameLoop

def main():
    display = pygame.display.set_mode((900,700))

    pygame.init()

    stage = Stage()
    running = True
    clock = Clock()
    clock.g_clock.tick(60)
    renderer = Renderer(stage, display)
    stage.add_new_block()
    start_time = 0
    game_loop = GameLoop(stage)
    speeds = [100, 600, 500, 450, 400, 350, 300, 250, 200, 150, 100]
    speed_level = (1,0)
    score = 0
    max_score = 0
    pygame.display.set_caption("Tetris")

    while running:

        game_loop.events()

        if game_loop.drop == True:
            speed = speeds[0]
        if game_loop.drop == False:
            speed = speeds[speed_level[0]]

        if (clock.get_ticks() - start_time) > speed:
            new_score = stage.drop_block()
            if new_score != None:
                score += new_score
            
            if (score - speed_level[1] > 100):
                if speed_level[0] < 10:
                    old = speed_level
                    speed_level = (old[0]+1,old[1]+100)

            start_time = clock.get_ticks()
        
        display.fill((0,0,0))
        renderer.render_info(score, max_score, speed_level[0])
        renderer.render_game()

        if stage.game_over:
            enterkey = False
            option = "yes"
            if score > max_score:
                max_score = score
            while True:
                renderer.game_over(option)
                input = game_loop.game_over()
                if input == "yes":
                    option = "yes"
                elif input == "no":
                    option = "no"
                elif input == True:
                    enterkey = True
                if option == "yes" and enterkey == True:
                    stage.game_over = False
                    break
                if option == "no" and enterkey == True:
                    running = False
                    break
            speed_level = (1,0)
            score = 0
            stage.retry()


    pygame.quit()

if __name__ == "__main__":
    main()